File: new_version/frontend/src/utils/dateTokens.ts
// Date token resolution shared between selectors and the public dashboard.
// The backend resolves the same tokens in `app/domain/services/date_tokens.py`,
// so the two sides MUST stay in sync.

export const DATE_TOKENS = [
  'TODAY',
  'YESTERDAY',
  'TOMORROW',
  'LAST_7_DAYS',
  'LAST_14_DAYS',
  'LAST_30_DAYS',
  'LAST_90_DAYS',
  'THIS_MONTH_START',
  'LAST_MONTH_START',
  'THIS_QUARTER_START',
  'LAST_QUARTER_START',
  'THIS_YEAR_START',
  'LAST_YEAR_START',
  'YEAR_START',
] as const

export type DateToken = (typeof DATE_TOKENS)[number]

const DATE_ONLY_RE = /^\d{4}-\d{2}-\d{2}$/

const pad = (n: number): string => String(n).padStart(2, '0')
const fmt = (d: Date): string =>
  `${d.getFullYear()}-${pad(d.getMonth() + 1)}-${pad(d.getDate())}`

function quarterStart(d: Date): Date {
  const month = Math.floor(d.getMonth() / 3) * 3
  return new Date(d.getFullYear(), month, 1)
}

export function isDateToken(value: unknown): value is DateToken {
  return typeof value === 'string' && (DATE_TOKENS as readonly string[]).includes(value)
}

export function isDateOnly(value: unknown): boolean {
  return typeof value === 'string' && DATE_ONLY_RE.test(value)
}

/**
 * Resolve a date token to a `YYYY-MM-DD` string. Pass-through for anything that
 * is not a recognized token.
 */
export function resolveDateToken(value: unknown): unknown {
  if (!isDateToken(value)) return value

  const now = new Date()

  switch (value) {
    case 'TODAY':
      return fmt(now)
    case 'YESTERDAY': {
      const d = new Date(now)
      d.setDate(d.getDate() - 1)
      return fmt(d)
    }
    case 'TOMORROW': {
      const d = new Date(now)
      d.setDate(d.getDate() + 1)
      return fmt(d)
    }
    case 'LAST_7_DAYS': {
      const d = new Date(now)
      d.setDate(d.getDate() - 7)
      return fmt(d)
    }
    case 'LAST_14_DAYS': {
      const d = new Date(now)
      d.setDate(d.getDate() - 14)
      return fmt(d)
    }
    case 'LAST_30_DAYS': {
      const d = new Date(now)
      d.setDate(d.getDate() - 30)
      return fmt(d)
    }
    case 'LAST_90_DAYS': {
      const d = new Date(now)
      d.setDate(d.getDate() - 90)
      return fmt(d)
    }
    case 'THIS_MONTH_START':
      return fmt(new Date(now.getFullYear(), now.getMonth(), 1))
    case 'LAST_MONTH_START':
      return fmt(new Date(now.getFullYear(), now.getMonth() - 1, 1))
    case 'THIS_QUARTER_START':
      return fmt(quarterStart(now))
    case 'LAST_QUARTER_START': {
      const thisQ = quarterStart(now)
      const prevQ = new Date(thisQ)
      prevQ.setMonth(prevQ.getMonth() - 3)
      return fmt(prevQ)
    }
    case 'THIS_YEAR_START':
    case 'YEAR_START':
      return fmt(new Date(now.getFullYear(), 0, 1))
    case 'LAST_YEAR_START':
      return fmt(new Date(now.getFullYear() - 1, 0, 1))
    default:
      return value
  }
}

/**
 * Resolve any date tokens hiding inside a filter value, mirroring the backend
 * `resolve_filter_value` shape: handles dicts (`{from, to}`), arrays, scalars.
 */
export function resolveFilterValue(value: unknown): unknown {
  if (value === null || value === undefined) return value
  if (Array.isArray(value)) return value.map(resolveDateToken)
  if (typeof value === 'object') {
    const out: Record<string, unknown> = {}
    for (const [k, v] of Object.entries(value as Record<string, unknown>)) {
      out[k] = resolveDateToken(v)
    }
    return out
  }
  return resolveDateToken(value)
}

/**
 * Human-readable label for a token (used in dropdowns and tooltips).
 */
export function tokenLabel(token: DateToken): string {
  switch (token) {
    case 'TODAY':
      return 'Сегодня'
    case 'YESTERDAY':
      return 'Вчера'
    case 'TOMORROW':
      return 'Завтра'
    case 'LAST_7_DAYS':
      return '7 дней назад'
    case 'LAST_14_DAYS':
      return '14 дней назад'
    case 'LAST_30_DAYS':
      return '30 дней назад'
    case 'LAST_90_DAYS':
      return '90 дней назад'
    case 'THIS_MONTH_START':
      return 'Начало месяца'
    case 'LAST_MONTH_START':
      return 'Начало прошлого месяца'
    case 'THIS_QUARTER_START':
      return 'Начало квартала'
    case 'LAST_QUARTER_START':
      return 'Начало прошлого квартала'
    case 'THIS_YEAR_START':
    case 'YEAR_START':
      return 'Начало года'
    case 'LAST_YEAR_START':
      return 'Начало прошлого года'
    default:
      return token
  }
}

File: backend/app/domain/services/date_tokens.py
"""Date token resolution for selector default values and filter values.

Tokens are dynamic placeholders like ``LAST_30_DAYS`` that resolve to actual
``YYYY-MM-DD`` strings at request time. This lets clients send a token in
``selector.config.default_value`` (or in raw filter values) instead of a fixed
date — the backend resolves it on every request, so the same default keeps
sliding forward as time passes.

Also provides ``extend_to_end_of_day`` so that ``BETWEEN`` filters with a
``to`` date like ``"2026-04-06"`` cover the full day instead of stopping at
``00:00:00``.
"""

from __future__ import annotations

import re
from datetime import date, datetime, timedelta
from typing import Any

# Token names — kept as a frozenset for fast membership checks.
DATE_TOKENS: frozenset[str] = frozenset({
    "TODAY",
    "YESTERDAY",
    "TOMORROW",
    "LAST_7_DAYS",
    "LAST_14_DAYS",
    "LAST_30_DAYS",
    "LAST_90_DAYS",
    "THIS_MONTH_START",
    "LAST_MONTH_START",
    "THIS_QUARTER_START",
    "LAST_QUARTER_START",
    "THIS_YEAR_START",
    "LAST_YEAR_START",
    "YEAR_START",  # alias for THIS_YEAR_START
})

_DATE_ONLY_RE = re.compile(r"^\d{4}-\d{2}-\d{2}$")


def _fmt(d: date) -> str:
    return d.strftime("%Y-%m-%d")


def _today() -> date:
    return datetime.now().date()


def _quarter_start(d: date) -> date:
    quarter_first_month = ((d.month - 1) // 3) * 3 + 1
    return date(d.year, quarter_first_month, 1)


def is_date_token(value: Any) -> bool:
    """Return True if value is a recognized date token string."""
    return isinstance(value, str) and value in DATE_TOKENS


def is_date_only(value: Any) -> bool:
    """Return True if value is a date-only string ``YYYY-MM-DD``."""
    return isinstance(value, str) and bool(_DATE_ONLY_RE.match(value))


def resolve_token(value: Any) -> Any:
    """Resolve a date token to ``YYYY-MM-DD``. Pass through other values.

    Args:
        value: Token name (``TODAY``, ``LAST_30_DAYS``, ...) or any other value.

    Returns:
        ``YYYY-MM-DD`` string for known tokens; the original value otherwise.
    """
    if not is_date_token(value):
        return value

    today = _today()

    if value == "TODAY":
        return _fmt(today)
    if value == "YESTERDAY":
        return _fmt(today - timedelta(days=1))
    if value == "TOMORROW":
        return _fmt(today + timedelta(days=1))
    if value == "LAST_7_DAYS":
        return _fmt(today - timedelta(days=7))
    if value == "LAST_14_DAYS":
        return _fmt(today - timedelta(days=14))
    if value == "LAST_30_DAYS":
        return _fmt(today - timedelta(days=30))
    if value == "LAST_90_DAYS":
        return _fmt(today - timedelta(days=90))
    if value == "THIS_MONTH_START":
        return _fmt(date(today.year, today.month, 1))
    if value == "LAST_MONTH_START":
        first_of_this_month = date(today.year, today.month, 1)
        last_month_last_day = first_of_this_month - timedelta(days=1)
        return _fmt(date(last_month_last_day.year, last_month_last_day.month, 1))
    if value == "THIS_QUARTER_START":
        return _fmt(_quarter_start(today))
    if value == "LAST_QUARTER_START":
        this_q = _quarter_start(today)
        last_q_last_day = this_q - timedelta(days=1)
        return _fmt(_quarter_start(last_q_last_day))
    if value in ("THIS_YEAR_START", "YEAR_START"):
        return _fmt(date(today.year, 1, 1))
    if value == "LAST_YEAR_START":
        return _fmt(date(today.year - 1, 1, 1))

    return value  # unreachable, but keeps type checkers happy


def resolve_filter_value(selector_type: str | None, value: Any) -> Any:
    """Resolve date tokens inside a filter value, respecting its shape.

    - For ``date_range``: ``{from, to}`` → resolves both keys.
    - For ``single_date`` or anything else: passes scalar through ``resolve_token``.
    - Lists are walked element-wise.
    """
    if value is None:
        return value
    if isinstance(value, dict):
        return {k: resolve_token(v) for k, v in value.items()}
    if isinstance(value, list):
        return [resolve_token(v) for v in value]
    return resolve_token(value)


def extend_to_end_of_day(value: Any) -> Any:
    """Append ``23:59:59`` to a date-only string so ``BETWEEN`` covers the day.

    Other values pass through unchanged.
    """
    if is_date_only(value):
        return f"{value} 23:59:59"
    return value
